src/contribute-to-tidb/contribute-code.md

```diff
@@ -1,88 +1,53 @@
+  5. When your pull request gets enough approvals(the default number is 2) and all other requirements are met, it will be merged.
+  6. Handle regressions introduced by your change. Although committers bear the main responsibility to fix regressions, it's quite nice for you to handle it(reverting the change or sending fixes).
 
-### 1. Create GitHub Issue and Reach Consensus
+Clear and kind communication is key to this process.
 
-The first step for making a contribution to TiDB is to reach consensus with the community. This means agreeing on the scope and implementation approach of a change.
+## Writing tests
 
-In most cases, the discussion should happen on [GitHub issues](https://github.com/pingcap/tidb/issues).
+One important thing when you make code contributions to TiDB is tests. Tests should be always considered as a part of your change. Any code changes that cause semantic changes or new function additions to TiDB should have corresponding test cases. And of course you can not break any existing test cases if they are still valid. It's recommended to [run tests](../get-started/write-and-run-unit-tests.md) on your local environment first to find obvious problems and fix them before opening the pull request.
 
-The following types of changes requires a dedicated topic on the [TiDB Internals forum](https://internals.tidb.io/):
+It's also highly appreciated if your pull request only contains test cases to increase test coverage of TiDB. Supplement test cases for existing modules is a good and easy way to become acquainted with existing code.
 
-* big changes (major new feature; big refactors, involving multiple components)
-* potentially controversial changes or issues
-* changes with very unclear approaches or multiple equal approaches
+## Making good pull requests
 
-Do not open a GitHub issue for these types of changes before the discussion has come to a conclusion. GitHub issue based on a forum discussion need to link to that discussion and should summarize the outcome.
+When creating a pull request for submission, there are several things that you should consider to help ensure that your pull request is accepted:
 
-Requirements for a GitHub issue to get consensus:
+  - Does the contribution alter the behavior of features or components in a way that it may break previous users’ programs and setups? If yes, there needs to be a discussion and agreement that this change is desirable.
+  - Does the contribution conceptually fit well into TiDB? Is it too much of a special case such that it makes things more complicated for the common case, or bloats the abstractions / APIs?
+  - Does the contribution make a big impact on TiDB's build time?
+  - Does your contribution affect any documentation? If yes, you should add/change proper documentation. 
+  - If there are any new dependencies, are they under active maintainences? What are their licenses?
```

Comment:
```suggestion
* Does the contribution alter the behavior of features or components in a way that it may break previous users' programs and setups? If yes, there needs to be a discussion and agreement that this change is desirable.
* Does the contribution conceptually fit well into TiDB? Is it too much of a special case such that it makes things more complicated for the common case, or bloats the abstractions/APIs?
* Does the contribution make a big impact on TiDB's build time?
* Does your contribution affect any documentation? If yes, you should add/change proper documentation. 
* If there are any new dependencies, are they under active maintenances? What are their licenses?
```